File: 2024/day04-2.ts
type Coords = {
  row: number;
  col: number;
};

function solution(content: string) {
  return content
    .split("\n")
    .filter((line) => line)
    .reduce(
      (rowAcc, row, rowIndex, matrix) => [
        ...rowAcc,
        ...row
          .split("")
          .reduce(
            (colAcc, col, colIndex) =>
              col === "A" && check(matrix, rowIndex, colIndex)
                ? [...colAcc, true]
                : colAcc,
            [] as boolean[]
          ),
      ],
      [] as boolean[]
    ).length;
}

function check(matrix: string[], row: number, col: number) {
  if (
    row === 0 ||
    row === matrix.length - 1 ||
    col === 0 ||
    col === matrix[0].length - 1
  )
    return false;
  return (
    (matrix[row - 1][col - 1] === "M" &&
      matrix[row + 1][col + 1] === "S" &&
      matrix[row + 1][col - 1] === "M" &&
      matrix[row - 1][col + 1] === "S") ||
    (matrix[row - 1][col - 1] === "M" &&
      matrix[row + 1][col + 1] === "S" &&
      matrix[row - 1][col + 1] === "M" &&
      matrix[row + 1][col - 1] === "S") ||
    (matrix[row + 1][col + 1] === "M" &&
      matrix[row - 1][col - 1] === "S" &&
      matrix[row - 1][col + 1] === "M" &&
      matrix[row + 1][col - 1] === "S") ||
    (matrix[row + 1][col + 1] === "M" &&
      matrix[row - 1][col - 1] === "S" &&
      matrix[row + 1][col - 1] === "M" &&
      matrix[row - 1][col + 1] === "S")
  );
}

Deno.readTextFile("day04.dat")
  .then((content) => content.trim())
  .then(solution)
  .then(console.log);

File: 2025/day04_2.py
# pylint: disable=missing-module-docstring, missing-function-docstring


from typing import Any

type Coords = tuple[int, int]


def solution(content: str) -> Any:
    return remove(parse(content))


def remove(rolls: frozenset[Coords]) -> int:
    to_remove = {roll for roll in rolls if is_accessible(rolls, roll)}
    if not to_remove:
        return 0
    return len(to_remove) + remove(
        frozenset(roll for roll in rolls if roll not in to_remove)
    )


def is_accessible(rolls: frozenset[Coords], roll: Coords) -> bool:
    row, col = roll
    return (
        len(
            [
                r
                for r in (
                    (row - 1, col - 1),
                    (row - 1, col),
                    (row - 1, col + 1),
                    (row, col - 1),
                    (row, col + 1),
                    (row + 1, col - 1),
                    (row + 1, col),
                    (row + 1, col + 1),
                )
                if r in rolls
            ]
        )
        < 4
    )


def parse(content: str) -> frozenset[Coords]:
    return frozenset(
        (row_index, col_index)
        for row_index, row in enumerate(content.split("\n"))
        for col_index, col in enumerate(row)
        if col == "@"
    )


def main() -> None:
    with open("day04.dat", "r", encoding="utf8") as file:
        print(solution(file.read().strip()))


if __name__ == "__main__":
    main()
